File: lintcode/python/0069_binary_tree_level_order_traversal.py
#!/usr/bin/python -t

# one queue with BFS

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

from collections import deque

class Solution:
    """
    @param root: A Tree
    @return: Level order a list of lists of integer
    """
    def levelOrder(self, root):
        # write your code here
        ret = []
        if root == None:
            return ret
            
        q = deque([root])
        
        while q:
            level = []
            for i in range(len(q)):
                node = q.popleft()
                level.append(node.val)
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            ret.append(level)
            
        return ret


# DFS solution


"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

from collections import deque

class Solution:
    """
    @param root: A Tree
    @return: Level order a list of lists of integer
    """
    def levelOrder(self, root):
        # write your code here
        ret = []
        if root == None:
            return ret
            
        q = deque([root])
        target_level = 0
        
        while True:
            level = []
            self.dfs(root, 0, target_level, level)
            if not level:
                break
            
            ret.append(level)
            target_level += 1
            
        return ret
        
    def dfs(self, root, cur_level, target_level, level):
        if root == None or cur_level > target_level:
            return
        
        if cur_level == target_level:
            level.append(root.val)
            
        self.dfs(root.left, cur_level+1, target_level, level)
        self.dfs(root.right, cur_level+1, target_level, level)

File: lintcode/golang/0069_binary_tree_level_order_traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

/**
 * @param root: A Tree
 * @return: Level order a list of lists of integer
 */
func levelOrder (root *TreeNode) [][]int {
    // write your code here
    ret := make([][]int, 0)
    if root == nil {
        return ret
    }
    
    q := make([]*TreeNode, 0)
    q = append(q, root)
    
    for len(q) != 0 {
        level := make([]*TreeNode, 0)
        tmp := make([]int, 0)
        for _, node := range q {
            tmp = append(tmp, node.Val)
            if node.Left != nil {
                level = append(level, node.Left)
            }
            if node.Right != nil {
                level = append(level, node.Right)
            }
        }
        q = level
        ret = append(ret, tmp)
    }
    
    return ret
}

